debian/mumbledj.1
.TH MUMBLEDJ 1 2016-02-23
.SH NAME
mumbledj \- a Mumble bot that plays music fetched from YouTube videos and SoundCloud tracks
.SH SYNOPSIS
.SY mumbledj
.OP \-server address
.OP \-port port
.OP \-username username
.OP \-password password
.OP \-channel channel
.OP \-cert pemCert
.OP \-key pemKey
.OP \-accesstokens accesstokens
.OP \-insecure
.SY
.SY mumbledj
.OP \-help
.YS
.SY mumbledj
.OP \-version
.YS
.SH OPTIONS
.TP
.BI \-server \ address
.TQ
.BI \-port \ port
Address and port number for Mumble server; default to
.B localhost
and
.BR 64738 ,
respectively.
.TP
.B \-username
.I username
Defaults to
.B MumbleDJ
.TP
.B \-password
.I password
The password for the Mumble server. Defaults to no password.
.TP
.B \-channel
.I channel
Channel joined by the bot. Defaults to the root channel.
.TP
.B \-cert
.I pemCert
.TQ
.B \-key
.I pemKey
Paths to files containing a client certificate and key in PEM format.
If unspecified, no client certificate will be presented.
.TP
.B \-accesstokens
.I accesstokens
List of access tokens for channel authentication.
Separated by spaces--take care to use correct quoting when launching
.B mumbledj
from a shell.
.TP
.B \-insecure
Causes the bot to not check the server certificate.
Try using this if you have having connection issues.
.TP
.B \-help
Prints a summary of options and then exits.
.TP
.B \-version
Prints the version number and then exits.
.SH SEE ALSO
Full documentation can be found at
.UR https://github.com/matthieugrieger/mumbledj
.UE ,
a copy of which can be found at /usr/share/doc/mumbledj/README.md.gz.
The file README.Debian in the same directory contains notes about how to configure systemd to pass the options above when running
.BR mumbledj .


CHANGELOG.md
MumbleDJ Changelog
==================

### February 14, 2016 -- `v2.8.15`
* Fixed an incorrectly formatted error message (thanks [@GabrielPlassard](https://github.com/GabrielPlassard)).

### February 12, 2016 -- `v2.8.14`
* Audio is now downloaded using the `bestaudio` format. This prevents situations in which some audio would not play because an `.m4a` file was not available (thanks [@mpacella88](https://github.com/mpacella88)).

### February 6, 2016 -- `v2.8.13`
* Added `!version` command to display the version of MumbleDJ that is running (thanks [@zeblau](https://github.com/zeblau)).
* Added `version` commandline argument to display the version of MumbleDJ that is running (thanks [@zeblau](https://github.com/zeblau)).

### January 26, 2016 -- `v2.8.12`
* Temporarily fixed discontinued code.google.com imports.

### January 14, 2016 -- `v2.8.11`
* Fixed: Unable to use offsets if it's formatted as &t vs ?t in the URL (thanks [@fiveofeight](https://github.com/fiveofeight)).

### January 11, 2016 -- `v2.8.10`
* Created a new configuration value in the General section called PlayerCommand. This allows the user to change between "ffmpeg" and "avconv" for playing audio files.
* Added check for valid PlayerCommand value. If the value is invalid the bot will default to `ffmpeg`.

### December 26, 2015 -- `v2.8.9`
* Fixed an incorrect `!currentsong` message for songs within playlists.

### December 21, 2015 -- `v2.8.8`
* Fixed a typo in song list HTML (thanks [@mkody](https://github.com/mkody)).

### December 19, 2015 -- `v2.8.7`
* Added AnnounceNewTracks config option (thanks [@HowIChrgeLazer](https://github.com/HowIChrgeLazer)).

### December 16, 2015 -- `v2.8.6`
* Added !addnext command (thanks [@nkhoit](https://github.com/nkhoit)).
* Added argument to !listsongs command to specify how many songs to list (thanks [@nkhoit](https://github.com/nkhoit)).

### December 14, 2015 -- `v2.8.5`
* Added !listsongs command (thanks [@nkhoit](https://github.com/nkhoit)).

### December 7, 2015 -- `v2.8.4`
* YouTube and SoundCloud API keys are now stored in the configuration file instead of environment variables. Existing installations with API keys in environment variables will automatically be migrated to the configuration file (thanks [@Gamah](https://github.com/Gamah)).

### October 16, 2015 -- `v2.8.3`
* Playlists can now be over 50 songs in length (thanks [@GabrielPlassard](https://github.com/GabrielPlassard)).
* Added MaxSongPerPlaylist configuration option.

### October 14, 2015 -- `v2.8.2`
* Fixed possible index out of range panic when auto shuffle is on (thanks [@GabrielPlassard](https://github.com/GabrielPlassard)).

### October 12, 2015 -- `v2.8.1`
* Added !shuffle, !shuffleon, and !shuffleoff commands (thanks [@GabrielPlassard](https://github.com/GabrielPlassard)).

### October 1, 2015 -- `v2.8.0`
* Added Soundcloud support (thanks [@MichaelOultram](https://github.com/MichaelOultram)).

### August 12, 2015 -- `v2.7.5`
* Fixed cache clearing earlier than expected (thanks [@CMahaff](https://github.com/CMahaff)).

### May 19, 2015 -- `v2.7.4`
* Fixed a panic that occurred when certain YouTube playlists were added to the queue.

### May 14, 2015 -- `v2.7.3`
* Fixed `!move` not working for subchannels (thanks [@mkbwong](https://github.com/mkbwong)).
* Fixed MumbleDJ showing invalid YouTube ID error message in chat when an invalid YouTube API key is supplied (thanks [@fiveofeight](https://github.com/fiveofeight)).
* Fixed MumbleDJ showing invalid YouTube ID error message in chat when a song exceeds the allowed time duration.

### May 12, 2015 -- `v2.7.2`
* Fixed incorrect values shown in timestamp for videos over an hour long.
* Reworked timestamp parsing.

### May 9, 2015 -- `v2.7.1`
* Added support for YouTube offsets. This means that YouTube URLs with the `t` parameter will start at the time specified in the URL instead of the beginning.
* Cleaned up comments in some files and removed some unnecessary code.
* Fixed a bug in which a duration of 0:00 was shown for songs that were less than a minute long.

### April 17, 2015 -- `v2.7.0`
* Migrated all YouTube API calls to YouTube Data API v3. This means that you **MUST** follow the instructions in the following link if you were using a previous version of MumbleDJ: https://github.com/matthieugrieger/mumbledj#youtube-api-keys.
* Made the SongQueue much more flexible. These changes will allow easy addition of support for other music services.

### March 28, 2015 -- `v2.6.10`
* Fixed a crash that would occur when the last song of a playlist was skipped.

### March 27, 2015 -- `v2.6.9`
* Fixed a race condition that would sometimes cause the bot to crash (thanks [dylanetaft](https://github.com/dylanetaft)!).

### March 26, 2015 -- `v2.6.8`
* Renamed `mumbledj.gcfg` to `config.gcfg`. However, please note that it will still be called `mumbledj.gcfg` in your `~/.mumbledj` directory. Hopefully this will avoid any ambiguity when referring to the
config files.
* Tweaked the `Makefile` to handle situations where `go build` creates an executable with an appended version number.

### March 20, 2015 -- `v2.6.7`
* Fixed a typo in `mumbledj.gcfg`.
* Songs and playlists are now skipped immediately if the submitter submits a skip command.
* `SONG_SKIPPED_HTML` and `PLAYLIST_SKIPPED_HTML` are no longer shown if the submitter or admin skips a song/playlist.

### March 7, 2015 -- `v2.6.6`
* Added missing AdminSkipPlaylistAlias option to `mumbledj.gcfg`.

### February 25, 2015 -- `v2.6.5`
* Added automatic connection retries if the bot loses connection to the server. The bot will attempt to reconnect to the server every 30 seconds for a period of 15 minutes, then exit if a connection cannot be made.

### February 20, 2015 -- `v2.6.4`
* Fixed failed audio downloads for YouTube videos with IDs beginning with "-".

### February 19, 2015 -- `v2.6.3`
* Added `gumbleutil.CertificateLockFile()` for more secure connections.

mumbledj/mumbledj.py
#----------------------#
#       MumbleDJ       #
#  By Matthieu Grieger #
#----------------------#---------------------------#
# mumbledj.py                                      #
# Contains definitions of MumbleDJ class & methods #
#--------------------------------------------------#

import pymumble
from pymumble.constants import *
from config import *
from time import sleep

class MumbleDJ:
	# Since all the configuration is set in config.py, we don't really need to do anything here.
	def __init__(self):
		print('Starting up ' + SERVER_USERNAME + '...') 
	
	# Connects to the Mumble server with the credentials specified upon object creation.	
	def connect_to_server(self):
		self.mumble = pymumble.Mumble(SERVER_ADDRESS, SERVER_PORT, SERVER_USERNAME, SERVER_PASSWORD, debug = DEBUG)
		self.mumble.callbacks.set_callback(PYMUMBLE_CLBK_TEXTMESSAGERECEIVED, self.parse_message)
		self.mumble.start()
		self.mumble.is_ready()
		self.mumble.channels.find_by_name(DEFAULT_CHANNEL).move_in()
		self.mumble.users.myself.mute()
		self.kill_bot = False
		
		self.listen()
	
	# Starts to play the first song in the queue when called. If no songs exist in the queue, it will wait until
	# a song is added.	
	def start_music(self):
		pass
		
	# Resumes music when called.
	def play_music(self):
		pass
		
	# Pauses music until told to resume.
	def pause_music(self):
		pass
		
	# Adds a YouTube link to the queue along with its metadata.
	def add_to_queue(self):
		pass
		
	# Sends a message to the chat when a new song starts playing.	
	def announce_new_song(self):
		pass
	
	# Skips the current song if more than 50% of the users in the channel vote to skip.
	# The vote tracking will be handled elsewhere.	
	def skip_song(self):
		pass
		
	# Raises the volume by the increment decided by the user. 
	def raise_volume(self, increment):
		pass
	
	# Lowers the volume by the increment decided by the user.	
	def lower_volume(self, increment):
		pass
		
	# Moves bot from the current channel to the specified channel.
	def move_bot(self, channel):
		pass
		
	# Completely stops the bot if a person on the approved list of admins
	# issues the stop command.
	def kill_bot(self):
		self.kill_bot = True
		
	# Parses a new message and checks for a command. If one exists, it will
	# parse the command and its arguments (if they exist) and pass them along
	# to the appropriate function.
	def parse_message(self, message):
		if message[0] == COMMAND_PREFIX:
			if ' ' in message:
				command = message[1:message.find(' ')]
			else:
				command = message[1:]
			
			if command == 'start' and ALLOW_START:
				pass
			elif command == 'play' and ALLOW_PLAY:
				pass
			elif command == 'pause' and ALLOW_PAUSE:
				pass
			elif command == 'add' and ALLOW_ADD:
				pass
			elif command == 'skip' and ALLOW_SKIPS:
				pass
			elif command == 'volumeup' and ALLOW_VOLUMEUP:
				pass
			elif command == 'volumedown' and ALLOW_VOLUMEDOWN:
				pass
			elif command == 'move' and ALLOW_MOVE:
				pass
			elif command == 'kill' and ALLOW_KILL:
				pass
		
	# This is the main loop for the bot. It will listen for commands periodically and
	# call the appropriate functions to deal with them.
	def listen(self):
		while self.mumble.is_alive() and not self.kill_bot:
			sleep(0.5)
	
	
	
